### PR TITLE
Introduce kvdb-lmdb

### PR DESCRIPTION
**Pull request by ordian**
On top of #120 (last commit only)

### CODE REVIEW

File: kvdb-lmdb/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+		Self::open(&Path::new(path), config.num_columns)
+	}
+}
+
+impl<'a> IterationHandler for &'a EnvironmentWithDatabases {
+	type Iterator = IterWithTxn<'a>;
+
+	fn iter(&self, col: usize) -> Self::Iterator {
+		// TODO: how to handle errors properly?
+		let ro_txn = self.ro_txn().expect("lmdb: transaction creation failed");
```
Comment (ordian) [inline kvdb-lmdb/src/lib.rs]: Read-only transaction creation can happen if too many thread are trying to create a read-only transaction or we're out of memory (cf. http://www.lmdb.tech/doc/group__mdb.html#gad7ea55da06b77513609efebd44b26920)

File: kvdb-lmdb/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+fn other_io_err<E>(e: E) -> io::Error where E: ToString {
+	io::Error::new(io::ErrorKind::Other, e.to_string())
+}
+
+type KeyValuePair = (Box<[u8]>, Box<[u8]>);
+
+/// Lmdb config.
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct DatabaseConfig {
+	num_columns: u32,
```
Comment (dvdplm) [inline kvdb-lmdb/src/lib.rs]: This field is private. Should it be public? Or have a constructor?

File: kvdb/src/lib.rs [hunk 3]
```diff
@@ -167,9 +181,425 @@ pub trait KeyValueDB: Sync + Send {
...
+							},
+						}
+					},
+				}
+			},
+			None => Ok(None),
+		}
+	}
+
+	/// Get value by partial key. Prefix size should match configured prefix size. Only searches flushed values.
```
Comment (dvdplm) [inline kvdb/src/lib.rs]:
```suggestion
	/// Get value by partial key. Prefix size should match configured prefix size. Only searches flushed values. Returns the first value with a matching key or `None`.
```

File: kvdb/src/lib.rs [hunk 4]
```diff
@@ -167,9 +181,425 @@ pub trait KeyValueDB: Sync + Send {
...
+	pub fn get_by_prefix(&self, col: Option<u32>, prefix: &[u8]) -> Option<Box<[u8]>> {
+		match self.iter_from_prefix(col, prefix).next() {
+			Some((k, v)) => if k.starts_with(prefix) { Some(v) } else { None },
```
Comment (dvdplm) [inline kvdb/src/lib.rs]: Why is this `starts_with` check necessary? Can LMDB return values that do not match the prefix?

File: kvdb-lmdb/src/lib.rs (hunk 2 above)
Comment (ordian) [reply kvdb-lmdb/src/lib.rs]: Fixed in 3e61712cc023bfb129e2cd1a5530148b80f4ca33

File: kvdb/src/lib.rs (hunk 4 above)
Comment (ordian) [reply kvdb/src/lib.rs]: It is actually coming from rocksdb implementation (https://github.com/paritytech/parity-common/blob/1409ddaf07584e0ddaf37c20fccc4a06ae8950fe/kvdb-rocksdb/src/lib.rs#L541).

File: kvdb/src/lib.rs [hunk 5]
```diff
@@ -167,9 +181,425 @@ pub trait KeyValueDB: Sync + Send {
...
+}
+
+
+
+impl<DB> DatabaseWithCache<DB>
+where
+	DB: OpenHandler<DB> + TransactionHandler,
+{
+	/// Opens or creates the database from the specified config and path.
+	pub fn open(config: &<DB as OpenHandler<DB>>::Config, path: &str) -> io::Result<Self> {
```
[the remark this answers is not in the document]
Comment (ordian) [reply kvdb/src/lib.rs]: We store this path as a String field [here](https://github.com/paritytech/parity-common/blob/1409ddaf07584e0ddaf37c20fccc4a06ae8950fe/kvdb-rocksdb/src/lib.rs#L541). Convertion from a Path to a String can fail (but not vice versa).

[the remark this answers is not in the document]
Comment (dvdplm) [reply kvdb/src/lib.rs]: That link is probably not what you meant!

File: kvdb/src/lib.rs (hunk 4 above)
Comment (dvdplm) [reply kvdb/src/lib.rs]: Interesting. I guess the same question applies! Doesn't rocksdb support prefix matching?

If it's a rocksdb quirk, maybe we can add a `TODO` to clean it up?

File: kvdb/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ordian) [reply kvdb/src/lib.rs]: Ah, sorry, I meant this one: https://github.com/paritytech/parity-common/pull/121/files#diff-7fe73f2989f300cb5d5dc23a9a54a212R245

File: kvdb/src/lib.rs (hunk 4 above)
Comment (ordian) [reply kvdb/src/lib.rs]: https://github.com/facebook/rocksdb/blob/120bc4715b7e851b32f9b880f5ffac1d4f3a5d50/include/rocksdb/options.h#L1148-L1154

https://github.com/paritytech/parity-common/blob/1409ddaf07584e0ddaf37c20fccc4a06ae8950fe/kvdb-rocksdb/src/lib.rs#L540

Comment (dvdplm) [reply kvdb/src/lib.rs]: For posterity, this seems to be fixed in [upstream rocksdb](https://docs.rs/rocksdb/0.12.1/rocksdb/struct.ReadOptions.html), from which we have diverged quite a bit I believe. Oddly enough the [parity-rocksdb fork](https://github.com/paritytech/rust-rocksdb/blob/master/README.md) names prefix search as one of the reasons for forking, and yet I can't find it in our code.

File: kvdb-lmdb/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+		assert!(db.drop_column().is_ok());
+		assert_eq!(db.num_columns(), 0);
+		// Don't drop the default column
+		assert!(db.drop_column().is_err());
+		assert_eq!(db.num_columns(), 0);
+	}
+
+	#[test]
+	fn test_create_path_if_missing() {
+		let tempdir = TempDir::new("test_crate_path_if_missing").unwrap();
```
Comment (dvdplm) [inline kvdb-lmdb/src/lib.rs]:
```suggestion
		let tempdir = TempDir::new("test_create_path_if_missing").unwrap();
```

File: kvdb-lmdb/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+		let config = DatabaseConfig {
```
Comment (dvdplm) [inline kvdb-lmdb/src/lib.rs]:
```suggestion
		let config = DatabaseConfig::new(1)
```

File: kvdb-lmdb/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+			num_columns: 1,
```
Comment (ordian) [inline kvdb-lmdb/src/lib.rs]:
```suggestion
```

File: kvdb-lmdb/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+		};
```
Comment (ordian) [inline kvdb-lmdb/src/lib.rs]:
```suggestion
```

**Comment by debris** [comment]
worth reading if we ever want to get back to this pr:
- https://www.reddit.com/r/rust/comments/dh96ta/zerocopylmdb_would_love_security_and_soundness

**Comment by ordian** [comment]
we're already using zerocopy bindings to lmdb here, but https://github.com/Kerollmops/heed also worth considering if we're getting back to this

**Comment by ordian** [comment]
Currently sled looks like a more promising approach and lmdb slow write speed is also something sled acknoledges https://github.com/spacejam/sled/wiki/motivating-experiences#lmdb

**Comment by ordian** [comment]
Closing as stale.

**Pull request closed without merging**